File: lc_soln/0074_search_a_2d_matrix/solution.cs
public class Solution
{
    public bool SearchMatrix(int[][] matrix, int target)
    {
        int nc = matrix[0].Length;

        int row = -1;
        int t = 0;
        int b = matrix.Length - 1;
        while (t <= b)
        {
            int mid = t + (b - t) / 2;
            if (matrix[mid][0] <= target && target <= matrix[mid][nc - 1])
            {
                row = mid;
                break;
            }
            if (matrix[mid][nc - 1] < target)
            {
                t = mid + 1;
            }
            else
            {
                b = mid - 1;
            }
        }

        if (row == -1)
        {
            return false;
        }

        int l = 0;
        int r = nc - 1;
        while (l <= r)
        {
            int mid = l + (r - l) / 2;
            if (matrix[row][mid] == target)
            {
                return true;
            }
            if (matrix[row][mid] < target)
            {
                l = mid + 1;
            }
            else
            {
                r = mid - 1;
            }
        }

        return false;
    }
}

File: lc_soln/0074_search_a_2d_matrix/solution.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        nr = len(matrix)
        nc = len(matrix[0])

        row = -1
        t = 0
        b = nr - 1
        while t <= b:
            mid = t + (b - t) // 2
            if matrix[mid][0] <= target <= matrix[mid][-1]:
                row = mid
                break
            if matrix[mid][-1] < target:
                t = mid + 1
            else:
                b = mid - 1

        if row == -1:
            return False

        l = 0
        r = nc - 1
        while l <= r:
            mid = l + (r - l) // 2
            if matrix[row][mid] == target:
                return True
            if matrix[row][mid] < target:
                l = mid + 1
            else:
                r = mid - 1

        return False


def main(): ...


if __name__ == "__main__":
    main()
